src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1040 @@
+    m = fromIntegral (maxBound :: Word32) :: Float
+{-# INLINE word32ToFloatInUnitInterval #-}
+
+-- The two integer functions below take an [inclusive,inclusive] range.
+randomIvalIntegral :: (RandomGen g, Integral a) => (a, a) -> g -> (a, g)
+randomIvalIntegral (l,h) = randomIvalInteger (toInteger l, toInteger h)
+
+{-# SPECIALIZE randomIvalInteger :: (Num a) =>
+    (Integer, Integer) -> StdGen -> (a, StdGen) #-}
+
+randomIvalInteger :: (RandomGen g, Num a) => (Integer, Integer) -> g -> (a, g)
+randomIvalInteger (l,h) rng
+ | l > h     = randomIvalInteger (h,l) rng
+ | otherwise = case f 1 0 rng of (v, rng') -> (fromInteger (l + v `mod` k), rng')
+     where
+       (genlo, genhi) = genRange rng
+       b = fromIntegral genhi - fromIntegral genlo + 1 :: Integer
+
+       -- Probabilities of the most likely and least likely result
+       -- will differ at most by a factor of (1 +- 1/q). Assuming the RandomGen
+       -- is uniform, of course
+
+       -- On average, log q / log b more pseudo-random values will be generated
+       -- than the minimum
+       q = 1000 :: Integer
+       k = h - l + 1
+       magtgt = k * q
+
+       -- generate pseudo-random values until we exceed the target magnitude
+       f mag v g | mag >= magtgt = (v, g)
+                 | otherwise = v' `seq`f (mag*b) v' g' where
+                        (x,g') = next g
+                        v' = v * b + (fromIntegral x - fromIntegral genlo)
+
+-- | Generate an integral in the range @[l, h]@ if @l <= h@ and @[h, l]@
+-- otherwise.
+uniformIntegralM :: (Bits a, Integral a, StatefulGen g m) => (a, a) -> g -> m a
+uniformIntegralM (l, h) gen = case l `compare` h of
+  LT -> do
+    let limit = h - l
+    let limitAsWord64 :: Word64 = fromIntegral limit
+    bounded <-
+      if fromIntegral limitAsWord64 == limit
+        -- Optimisation: if 'limit' fits into 'Word64', generate a bounded
+        -- 'Word64' and then convert to 'Integer'
+        then fromIntegral <$> unsignedBitmaskWithRejectionM uniformWord64 limitAsWord64 gen
+        else boundedExclusiveIntegralM (limit + 1) gen
+    return $ l + bounded
+  GT -> uniformIntegralM (h, l) gen
+  EQ -> pure l
+{-# INLINEABLE uniformIntegralM #-}
+
+-- | Generate an integral in the range @[0, s)@ using a variant of Lemire's
+-- multiplication method.
+--
+-- Daniel Lemire. 2019. Fast Random Integer Generation in an Interval. In ACM
+-- Transactions on Modeling and Computer Simulation
+-- https://doi.org/10.1145/3230636
+--
+-- PRECONDITION (unchecked): s > 0
+boundedExclusiveIntegralM :: forall a g m . (Bits a, Integral a, StatefulGen g m) => a -> g -> m a
+boundedExclusiveIntegralM (s :: a) gen = go
+  where
+    n = integralWordSize s
+    -- We renamed 'L' from the paper to 'k' here because 'L' is not a valid
+    -- variable name in Haskell and 'l' is already used in the algorithm.
+    k = wordSizeInBits * n
+    twoToK = (1::a) `shiftL` k
+    modTwoToKMask = twoToK - 1
+
+    t = (twoToK - s) `mod` s
```

Comment:
Since both arguments are positive, it is [better](http://hackage.haskell.org/package/integer-gmp-1.0.3.0/docs/src/GHC.Integer.Type.html#line-786) to use `rem` instead of `mod`.

Reply:
I am not deeply familiar with this algorithm. @curiousleo Do we know for sure that `twoToK >= s`?

Reply:
Yes: e.g. on a 64-bit machine, `k` is the number of bits in `s`, rounded up to the nearest word size. So `2^k = twoToK >= s`.

Reply:
Got it. Optimization applied